Projectorville

michaelpelletier opened the pull request:
Will close https://github.com/MegaGameSociety/WatchTheSkiesDashboard/issues/29 when it is done.

First looked like this:

<img width="1181" alt="screen shot 2016-01-17 at 9 55 43 pm" src="https://cloud.githubusercontent.com/assets/1640601/12382173/2107edee-bd65-11e5-84b7-ff88a14ddbb4.png">

Then a style overhaul happened when I had the idea to make it look like a News Broadcast: 

<img width="1653" alt="screen shot 2016-01-20 at 10 08 53 pm" src="https://cloud.githubusercontent.com/assets/1640601/12470616/a254a210-bfc3-11e5-8c20-4bf311c91674.png">

And continuing with progress.

<img width="837" alt="screen shot 2016-02-06 at 2 23 15 pm" src="https://cloud.githubusercontent.com/assets/1640601/12873881/b860fe56-cd96-11e5-9129-c00817764fff.png">

michaelpelletier commented on app/controllers/api/api_controller.rb:
```diff
@@ -26,22 +26,33 @@ def dashboard
+          "change" => countries_income_change[country]
+        }
+
+        @countries_data.push(this_country)
```
Been a long damn time since I've written Ruby and I was never particularly versed in it so I'm not super confident on this being the best way to do this. It also no doubt has broken some tests that I still need to address.

michaelpelletier commented on app/controllers/api/api_controller.rb:
```diff
@@ -25,23 +25,25 @@ def dashboard
     begin
-      @countries_data = {}
-      countries_calculations = {}
+      @countries_data = Array.new
```
I had a comment in here before about this being gross, which was lost in a refactor. This is probably still on the gross side, and I also need to remember to fix the tests for it.

michaelpelletier commented on app/assets/javascripts/angular/controllers/dashboard.js:
```diff
@@ -12,30 +12,14 @@ dashboardController.controller('DashboardCtrl', ['$rootScope', '$scope', '$http'
-          $(".Vatican").removeClass("alien");
-        }
+        $('.body').toggleClass('alien', result['alien_comms']);
+        $('.Vatican').toggleClass('alien', result['vatican_alien_comms']);
```
I didn't remove it since I wasn't sure if it was coming later, but fyi there is currently no class `".Vatican"` used anywhere in the app except for right here. This will never trigger.

FifthSurprise commented:
@michaelpelletier Just a reminder to give me a ping on either on Slack or on Github when you are done with this so I can review.

michaelpelletier commented:
@FifthSurprise Will do! You can always start reviewing it early if you would like.

michaelpelletier commented on app/views/layouts/application.html.erb:
```diff
@@ -3,10 +3,15 @@
     <title>WatchTheSkiesDashboard</title>
     <%= stylesheet_link_tag    'application', media: 'all'%>
+
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.4.3/angular.js"></script>
```
Note to self: Figure out why Bower stopped playing nicely with Angular. 
It's probably fine to use this CDN approach, but if that's the case, let's rip the logic out elsewhere.

FifthSurprise commented:
Not sure why but it looks like you can lose countries from the dashboard.  What I think is happening is on page load, you try to fit all of the tickers into one line but if there's no space, additional countries get pushed below visible window height and then do not appear even after manipulating resizing the window.

FifthSurprise commented on app/views/games/dashboard.html.erb:
```diff
@@ -1,111 +1,90 @@
+    <ul>
+      <li>
+        <div ng-repeat="country in countries" class="country">
+          <span>{{country.name}}</span>
```
I'm not sure if I like Country Name, Amount, then Glyphicon.  I think maybe a | as a delimiter would be a good idea between countries.

michaelpelletier replied:
Sure, easy fix.

FifthSurprise commented on app/views/games/dashboard.html.erb:
```diff
@@ -1,111 +1,90 @@
 
-      <div class="news-container active" ng-if="news.length==0">
-        <h4>Nothing has happened yet<h4>
+    <div id="news-image" ng-class="news[newsIndex].media_landscape == true ? 'landscape' : 'portrait'" ng-if="news[newsIndex].media_url">
```
If we have padding vertically, we should just shrink the image viewer to fit instead of showing the padding.  Unless this is really difficult, then I'd just settle for centering.

michaelpelletier replied:
I don't think there's padding being added now. The height is actually calculated by the size of the horizontal image. The difference here is because the actual _position_ of the image changes based on the orientation.

FifthSurprise commented on db/schema.rb:
```diff
@@ -57,6 +56,7 @@
     t.string   "media_url"
     t.boolean  "visible_content"
     t.boolean  "visible_image"
+    t.boolean  "media_landscape", default: false
```
I'm assuming we stored this because you want me to have some control somewhere about whether a media image should be larger landscape wise or not?

Can you open up an issue for all the things you want backend to track about an incoming image? I'm guessing if we get an image from twitter or somewhere, we should do some image processing to convert it into whatever you need?  Or can we just use image width/length?  Regardless, let's document this somewhere in an open issue so someone can work on it.

michaelpelletier replied:
I did! https://github.com/MegaGameSociety/WatchTheSkiesDashboard/issues/39

FifthSurprise replied:
Ah, I misunderstood the PR title.  Thought it was for news messages/ news paper.  Will update the issue to clarify.

FifthSurprise commented:
Oversight on my part but let's update the README to also `bower install`

FifthSurprise commented:
I discovered a bug in the api_controller.rb that was affecting you.

`income_change = if previous_income.empty?`
should be:
`income_change = if previous_income.nil? or previous_income.empty?`

I will make this change in PR #40  but you can also just put in the above change if you need it.  It handles not being able to compare previous income in the opening round against current income since there is no previous income.

michaelpelletier commented:
@FifthSurprise Yeah I need to fix the bower install thing and figure out why it's being an asshole with Angular. And good catch with the `nil` thing!

FifthSurprise commented on app/assets/javascripts/angular/controllers/dashboard.js:
```diff
@@ -49,30 +38,110 @@ dashboardController.controller('DashboardCtrl', ['$rootScope', '$scope', '$http'
     }, 3000);
   }
-]);
+]).animation('.slide-animation', function () {
```
I'm guessing this stuff is for animating left to right scroll for the ticker?  Some comments could be useful here to that effect.

michaelpelletier replied:
It's for the news itself, not the stock ticker thing, which definitely means it needs comments.

FifthSurprise commented:
@michaelpelletier Done looking at this for now.

I'd be interested in maybe throwing a small error image somewhere if the API call gets something other than a 200 response just to make it obvious that something weird happened with the API call.  Currently, the dashboard API call will have errors but still try to deliver what it can.  So you can use that 200 status message as a healthcheck and just display somewhere on the dashboard a small icon if something is going wrong.  Or I guess, throw a javascript error if it sees a 500 but that seems more painful.

Otherwise, the ticker resize thing seems to be the only weirdness.

FifthSurprise commented on app/controllers/api/api_controller.rb:
```diff
@@ -25,23 +25,23 @@ def dashboard
-      unless previous_income_list.empty?
-        previous_income_list.each do |pair|
-          countries_calculations[pair[0]] -= pair[1]
+        income_change = if previous_income.empty?
```
Derp, realized I can't fix this in #40 since bug is right here.

Anyway, please fix haha.

michaelpelletier commented:
Left to do:

- [x] Rebase and fix merge conflicts.
- [ ] Add small icon somewhere for if the API explodes.
- [ ] Try to get Angular working pleasantly through Bower again. 
- [ ] Add more comments in the JS
- [ ] Add | between countries on the news ticker.
- [ ] Fix weirdness with ticker sizing / resizing.

michaelpelletier commented:
@FifthSurprise I think this is ready for review now.

I ended up not being able to get `bower` to play nicely with Angular and Angular-Animate, and per our discussion previously, I don't think a CDN is a good way to go either, so what I did was create a `lib` folder inside the JS folder and just include the files that way. With Angular's upgrade path to 2.0 being risky (at best), we wouldn't simply want to use Bower to handle any such upgrade anyway (nor have it be a "oh let's just do it" thing). This method also allowed me to easily specify that I wanted to specifically use the minified versions, so we should get a nice little performance gain there, too!

I added a bunch more comments to the JS, which I hope clear some things up. The `dashboard.js` file has been almost entirely rewritten, so please do review it and make sure that everything makes sense. Likewise, I think you saw this already, but I did modify the `api_controller` to give us back some more information about the countries income. It's been ~3 years since I've written _any_ Ruby on Rails and I was never that proficient in the first place, so this could use a good review, too.

FINALLY, I adjusted the `get` request within the dashboard.js file to have a degree of error handling, so there's now an icon on the dashboard for if there happens to be a connection problem. It may well be a little bit large, but I'm not sure how close you guys typically are to the projector itself, and I wanted to make sure that it would be visible. 

<img width="795" alt="screen shot 2016-04-02 at 10 51 41 am" src="https://cloud.githubusercontent.com/assets/1640601/14227024/f3e4753e-f8c0-11e5-96d6-be43884199f7.png">

FifthSurprise commented:
Initial look seems like everything is fine.  I am going to work on automated testing on rails side in the nearish future so hopefully that will be less of a concern.  Also the icon is perfect.

I'm going to poke around a bit more to see if there's anything weird I don't understand but will merge it in soonish.

michaelpelletier commented:
Yay! :smile:
